src/skvhttpclient/src/skvhttp/common/StreamBuf.h

```diff
@@ -0,0 +1,94 @@
+/*
+MIT License
+
+Copyright(c) 2020 Futurewei Cloud
+
+    Permission is hereby granted,
+    free of charge, to any person obtaining a copy of this software and associated documentation files(the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and / or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions :
+
+    The above copyright notice and this permission notice shall be included in all copies
+    or
+    substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS",
+    WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+    DAMAGES OR OTHER
+    LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+
+#pragma once
+#include <streambuf>
+
+#include <skvhttp/common/Binary.h>
+#include <skvhttp/common/Common.h>
+
+namespace skv::http {
+
+template<typename WriterT, size_t bufSize>
+class WriteStreamBuf : public std::streambuf {
+public:
+    WriteStreamBuf(WriterT& writer): _writer(writer) {}
+
+    std::streamsize xsputn(const char_type* s, std::streamsize count) {
+        if (_sz + count > bufSize) {
+            throw std::out_of_range("not enough memory put");
+        }
+        std::memcpy(_data + _sz, s, count);
+        _sz += count;
+        return count;
+    }
+
+    int_type sputc(char_type ch) {
+        if (_sz + 1 > bufSize) {
+            throw std::out_of_range("not enough memory put");
+        }
+        _data[_sz] = ch;
+        ++_sz;
+        return ch;
+    }
+
+    ~WriteStreamBuf() {
+        Binary bin(_data, _sz, [](){});
+        _writer.write(bin);
+    }
+private:
+    WriterT& _writer;
+    char _data[bufSize];
```

Comment:
Can we make this a vector that we call "resize(bufSize)" on and resize if needed? As it is we have a very complicated serialization scheme that still relies on some underlying assumptions about the decimal types that we don't have control over.

Reply:
Thanks for the review! I considered using vector before, but Ivan told me to not to use it for efficiency reasons.
My concern is that vector may not copy bytes in bulk. If it always read/write individual bytes, the performance may be poor.
Btw, we are not assuming that the serialized size is exactly bufSize bytes. We are just allocating enough bytes for the serialized size. Could we just use a larger number for bufSize to be safe?
FYI:
Sizeof(cpp_dec_float_50) = 56; and the serialized size is 63 bytes.
Sizeof(cpp_dec_float_100) = 80; and the serialized size is 87 bytes.

Reply:
The reason I didn't want a vector is that vector will allocate for each decimal that is being serialized. With this approach here, we are using a stack variable which should be considerably faster (considering the small data size) as it doesn't involve the allocator and can be optimized and cached better.

This isn't decimal specific - it is a StreamBuffer implementation which is targeting small data of some consistent size. The assumption i guess is that you should use this class only for serializing data types of small and very consistent size.

For safety, the code is throwing out_of_range exceptions when there is an attempt to store more data than buffer size. I don't think any additional safety is needed.